[17.0][MIG] sale_rental

edescalona opened the pull request:
@BinhexTeam 

Some minor changes were made to it.

Commit:
Port sale_start_end_dates and sale_rental (#353)

dvdhinesh committed:
sale_start_end_dates and sale_rental

alexis-via committed:
Product qty on rental pickings

Replace openerp by odoo in import declarations

alexis-via committed:
Auto configure sale_rental module via post_install.py script

alexis-via committed:
sale_rental: fix warning @constrains parameter 'must_have_dates' is not writeable

rousseldenis committed:
pylint

Commit:
sale_rental: Migration to 11.0

alexis-via committed:
[MIG] sale_start_end_dates and sale_rental to v12

Convert readme to new subdir format

alexis-via committed:
sale_rental: Improve the code of object sale.rental

gurneyalex committed:
[IMP] enhancements suggested in code reviews

the creation of the stock location for rental respects the translations

sadamo committed:
[FIX] Archive rental routes rather than delete

oca-travis committed:
[UPD] Update sale_rental.pot

OCA-git-bot committed:
[UPD] README.rst

yweng8111 committed:
[IMP] sale_rental: extract function _run_rental_procurement

OCA-git-bot committed:
sale_rental 12.0.1.1.0

alexis-via committed:
[MIG] sale_rental from v12 to v14

No more need to add users to the packaging group
View of rental group is now possible on product.template form view

alexis-via committed:
sale_rental: black, isort, etc.

OCA-git-bot committed:
[UPD] README.rst

kirca committed:
[MIG] sale_rental: Migration to 15.0

Standard migration + add some tests

kirca committed:
[IMP] : black, isort, prettier, pylint

Commit:
[UPD] Update sale_rental.pot

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
sale_rental 15.0.1.0.1

Murtaza-OSI committed:
[16.0][MIG] sale_rental

Commit:
[IMP] sale_rental: black, isort, prettier

Commit:
[FIX] rework unit test

Commit:
[IMP] sale_rental: black, isort, prettier

rousseldenis commented:
/ocabot migration sale_rental

remi-filament commented on sale_rental/views/product.xml:
```diff
@@ -0,0 +1,74 @@
+        <field name="inherit_id" ref="product.product_normal_form_view" />
+        <field name="arch" type="xml">
+            <group name="sale" position="after">
+                <group name="rental" string="Rental" invisible="type != 'service'">
```
```suggestion
                <group name="rental" string="Rental" invisible="detailed_type != 'service'">
```
You should use the new field detailed_type here I suppose since type is computed from this new field.

remi-filament commented on sale_rental/views/product.xml:
```diff
@@ -0,0 +1,74 @@
+        <field name="model">product.template</field>
+        <field name="inherit_id" ref="product.product_template_only_form_view" />
+        <field name="arch" type="xml">
+            <group name="sale" position="inside">
+                <label
+                    for="rented_product_tmpl_id"
+                    string="Rental"
+                    invisible="type != 'service' or product_variant_count &gt; 1"
+                />
+                <field
+                    name="rented_product_tmpl_id"
+                    invisible="type != 'service' or product_variant_count &gt; 1"
+                />
+                <label
+                    for="rental_service_tmpl_ids"
+                    string="Rental Services"
+                    invisible="type == 'service' or product_variant_count &gt; 1"
+                />
+                <field
+                    name="rental_service_tmpl_ids"
+                    nolabel="1"
+                    invisible="rental_service_tmpl_ids == []"
+                />
+                <button
+                    type="action"
+                    name="%(sale_rental.create_rental_product_action)d"
+                    string="Create Rental Service"
+                    invisible="type == 'service' or product_variant_count &gt; 1"
+                />
+            </group>
```
The view is not so visible, I would suggest the following instead:
```suggestion
            <page name="sales" position="after">
                <page name="rental" string="Rental" invisible="not sale_ok and detailed_type=='service'">
                    <group>
                        <group name="rental-service" invisible="detailed_type != 'service' or product_variant_count &gt; 1">
                            <field name="rented_product_tmpl_id" />
                        </group>
                        <group name="rental-product" string="Rental Services" col="1" colspan="3"
                                invisible="detailed_type == 'service' or product_variant_count &gt; 1">
                            <field
                                nolabel="1"
                                name="rental_service_tmpl_ids"
                                invisible="not rental_service_tmpl_ids"
                            >
                                <tree create="0" delete="0">
                                    <field name="name" />
                                    <field name="list_price" string="Rental Price per Day"/>
                                </tree>
                            </field>
                            <button
                                type="action"
                                name="%(sale_rental.create_rental_product_action)d"
                                string="Create Rental Service"
                                colspan="3"
                            />
                        </group>
                    </group>
                </page>
            </page>
```

remi-filament commented on sale_rental/views/sale_rental.xml:
```diff
@@ -0,0 +1,193 @@
+        id="sale_rental_menu"
+        parent="sale.sale_order_menu"
+        action="sale_rental_action"
+        sequence="2"
```
```suggestion
        sequence="25"
```
The menu is shown as first on sale which does not seem relevant, better to have it later in the menu list.

remi-filament reviewed (changes requested):
Thanks @edescalona for porting this module.
I discovered it with your PR, it seems to be working fine.
However, display on products and sale order could be improved !

edescalona commented:
> Thanks @edescalona for porting this module. I discovered it with your PR, it seems to be working fine. However, display on products and sale order could be improved !

Hi @remi-filament thank you for your comments, I will review them and comment on them, greetings

Commit:
[UPD] Update sale_rental.pot

OCA-git-bot committed:
[BOT] post-merge updates

remi-filament commented on sale_rental/views/sale_rental.xml:
```diff
@@ -0,0 +1,193 @@
+        parent="sale.sale_order_menu"
+        action="sale_rental_action"
+        sequence="2"
+    />
```
Also, when you do not have sales permission, you still get the Sales menu with only Rental inside. I guess it is better to keep the same rights as for other menus so that Sales is not shown at all if user does not have sales rights.
```suggestion
        groups="sales_team.group_sale_salesman"
    />
```

edescalona commented:
Hello @remi-filament , ready for review, thanks.

remi-filament commented on sale_rental/__manifest__.py:
```diff
@@ -0,0 +1,28 @@
+    "author": "Akretion, Sodexis, Odoo Community Association (OCA)",
+    "maintainers": ["alexis-via"],
+    "website": "https://github.com/OCA/vertical-rental",
+    "depends": ["sale_start_end_dates", "sale_stock", "sales_team"],
```
sales_team is not needed here, since both other modules depends on sale which in turn depends on sales_team.
On the other side, I think sale_management should be added here as a dependency, since if not installed you will not get the Sales menu shown so you cannot access any of the functionality added by this module.
```suggestion
    "depends": ["sale_start_end_dates", "sale_stock", "sale_management"],
```

remi-filament commented on sale_rental/README.rst:
```diff
@@ -0,0 +1,136 @@
+Configuration
+=============
+
+In the *Products* menu, on the form view of a stockable product or
```
```suggestion
In the *Products* menu, on the form view of a storable product or
```

remi-filament commented on sale_rental/data/rental_data.xml:
```diff
@@ -0,0 +1,26 @@
+    <field name="sequence">100</field>
+    <field name="warehouse_selectable" eval="True" />
+    <field name="product_selectable" eval="False" />
+</record>
```
```suggestion
    <field name="company_id" eval="False" />
</record>
```
As per https://github.com/odoo/odoo/blob/a301f8ae6aad59efd073a34c6f9ec323191d211b/addons/stock/models/stock_location.py#L477 company will be set if not forced to False.
Here we want to keep the routes without company so that they can be reused on other companies.

remi-filament commented on sale_rental/data/rental_data.xml:
```diff
@@ -0,0 +1,26 @@
+<record id="route_warehouse0_sell_rented_product" model="stock.route">
+    <field name="name">Sell Rented Product</field>
+    <field name="sequence">100</field>
+</record>
```
```suggestion
    <field name="company_id" eval="False" />
</record>
```
Same applies here.

remi-filament commented on sale_rental/data/rental_data.xml:
```diff
@@ -0,0 +1,26 @@
+    <field name="sequence">100</field>
+</record>
+
+<record id="stock.warehouse0" model="stock.warehouse">
+    <field name="rental_allowed" eval="True" />
+</record>
```
I do not think we should do that in data, maybe in demo data rather.
When installing the module on existing DB with multiple companies or warehouses, we do not know on which one they want to allow rental, so it is probably better if this is done manually and added to configuration part of readme.

remi-filament commented on sale_rental/models/stock_warehouse.py:
```diff
@@ -0,0 +1,251 @@
+                    {
+                        "route_ids": [(4, rental_route.id)],
+                        "rental_route_id": rental_route.id,
+                        "sell_rented_product_route_id": sell_rented_route.id,
+                    }
+                )
+                rental_rules = self.env["stock.rule"].search(
+                    [
+                        ("route_id", "in", [rental_route.id, sell_rented_route.id]),
+                        ("active", "=", False),
+                    ]
+                )
+                if rental_rules:
+                    rental_rules.write({"active": True})
+                else:
+                    for rule_vals in self._get_rental_push_pull_rules():
+                        self.env["stock.rule"].create(rule_vals)
+            else:
+                for wh in self:
+                    rules_to_archive = self.env["stock.rule"].search(
+                        [
+                            (
+                                "route_id",
+                                "in",
+                                (
+                                    wh.rental_route_id.id,
+                                    wh.sell_rented_product_route_id.id,
+                                ),
+                            ),
+                            ("company_id", "=", wh.company_id.id),
+                        ]
+                    )
+                    rules_to_archive.write({"active": False})
+                    wh.write(
+                        {
+                            "route_ids": [(3, rental_route.id)],
+                            "rental_route_id": False,
+                            "sell_rented_product_route_id": False,
+                        }
+                    )
```
This should be done in function create as well.
Otherwise if you create a new warehouse, setting rental_allowed to true the locations / rules will not be created and routes will not be assigned.

remi-filament commented on sale_rental/views/product.xml:
```diff
@@ -0,0 +1,78 @@
+        <field name="model">product.product</field>
+        <field name="inherit_id" ref="product.product_normal_form_view" />
+        <field name="arch" type="xml">
+            <group name="sale" position="after">
+                <group
+                    name="rental"
+                    string="Rental"
+                    invisible="detailed_type != 'service'"
+                >
+                    <field name="rented_product_id" />
+                </group>
+                <group
+                    string="Rental Services"
+                    name="rental_services"
+                    invisible="detailed_type == 'service'"
+                >
+                    <button
+                        colspan="2"
+                        type="action"
+                        name="%(sale_rental.create_rental_product_action)d"
+                        string="Create Rental Service"
+                    />
+                    <field
+                        colspan="2"
+                        name="rental_service_ids"
+                        nolabel="1"
+                        readonly="1"
+                        invisible="rental_service_ids == []"
+                    />
+                </group>
+            </group>
```
I suggest to add a page specific for Rental and to add cols and colspan for avoiding getting a full width button.

```suggestion
            <page name="sales" position="after">
                <page
                    name="rental"
                    string="Rental"
                    invisible="not sale_ok and detailed_type=='service'"
                >
                    <group>
                        <group
                            name="rental_product"
                            invisible="detailed_type != 'service'"
                        >
                            <field name="rented_product_id" />
                        </group>
                        <group
                            name="rental_services"
                            string="Rental Services"
                            invisible="detailed_type == 'service'"
                            col="1"
                            colspan="3"
                        >
                            <field
                                colspan="2"
                                name="rental_service_ids"
                                nolabel="1"
                                readonly="1"
                                invisible="rental_service_ids == []"
                            />
                            <button
                                colspan="2"
                                type="action"
                                name="%(sale_rental.create_rental_product_action)d"
                                string="Create Rental Service"
                            />
                        </group>
                    </group>
                </page>
            </page>
```

remi-filament commented on sale_rental/views/product.xml:
```diff
@@ -0,0 +1,78 @@
+        <field name="model">product.template</field>
+        <field name="inherit_id" ref="product.product_template_only_form_view" />
+        <field name="arch" type="xml">
+            <group name="sale" position="inside">
+                <label
+                    for="rented_product_tmpl_id"
+                    string="Rental"
+                    invisible="detailed_type != 'service' or product_variant_count &gt; 1"
+                />
+                <field
+                    name="rented_product_tmpl_id"
+                    invisible="detailed_type != 'service' or product_variant_count &gt; 1"
+                />
+                <label
+                    for="rental_service_tmpl_ids"
+                    string="Rental Services"
+                    invisible="detailed_type == 'service' or product_variant_count &gt; 1"
+                />
+                <field
+                    name="rental_service_tmpl_ids"
+                    nolabel="1"
+                    invisible="rental_service_tmpl_ids == []"
+                />
+                <button
+                    type="action"
+                    name="%(sale_rental.create_rental_product_action)d"
+                    string="Create Rental Service"
+                    invisible="detailed_type == 'service' or product_variant_count &gt; 1"
+                />
+            </group>
```
Same here : 
```suggestion
            <page name="sales" position="after">
                <page
                    name="rental"
                    string="Rental"
                    invisible="not sale_ok and detailed_type=='service'"
                >
                    <group>
                        <group
                            name="rental-product"
                            invisible="detailed_type != 'service' or product_variant_count &gt; 1"
                        >
                            <field name="rented_product_tmpl_id" />
                        </group>
                        <group
                            name="rental-services"
                            string="Rental Services"
                            col="1"
                            colspan="3"
                            invisible="detailed_type == 'service' or product_variant_count &gt; 1"
                        >
                            <field
                                nolabel="1"
                                name="rental_service_tmpl_ids"
                                invisible="not rental_service_tmpl_ids"
                            >
                                <tree create="0" delete="0">
                                    <field name="name" />
                                    <field
                                        name="list_price"
                                        string="Rental Price per Day"
                                    />
                                </tree>
                            </field>
                            <button
                                type="action"
                                name="%(sale_rental.create_rental_product_action)d"
                                string="Create Rental Service"
                                colspan="3"
                            />
                        </group>
                    </group>
                </page>
            </page>
```

remi-filament reviewed (changes requested):
Thanks for your work on that module !

I made a first review, and I think there are some improvements to be made to that module, I made some suggestions inline (for aligning UI with the rest of Odoo interface, for handling multi-company properly), documentation also would need to be updated.

Regarding PR process, you need to clean up bot commits as specified here : https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-17.0 : 
"Squash administrative commits (if any) with the previous commit for reducing commit noise. Check https://github.com/OCA/maintainer-tools/wiki/Merge-commits-in-pull-requests#mergesquash-the-commits-generated-by-bots-or-weblate for details."

edescalona commented:
Hi @remi-filament , I'll take your comments into account, but this is a PR for the migration itself. It would be better to add the improvements you mentioned to a separate PR so as not to change the original format. Thanks for your feedback.

chaule97 reviewed (approved):
LGTM

edescalona committed:
[IMP] sale_rental: pre-commit auto fixes

edescalona committed:
[MIG] sale_rental: Migration to 17.0

edescalona commented:
Hi @remi-filament , I've corrected some of your suggestions. I think the rest should be addressed in a pull request, unless @rousseldenis  thinks they can be done in this same pull request? I also cleaned up the commit history a bit. Let me know what you think. Thanks.

PicchiSeba commented on sale_rental/data/rental_data.xml:
```diff
@@ -5,18 +5,20 @@
     <field name="sequence">100</field>
     <field name="warehouse_selectable" eval="True" />
     <field name="product_selectable" eval="False" />
+    <field name="company_id" eval="False" />
```
**question**: why is this needed? Is it to make it available to other companies as well?

PicchiSeba commented on sale_rental/data/rental_data.xml:
```diff
@@ -5,18 +5,20 @@
 <record id="route_warehouse0_sell_rented_product" model="stock.route">
     <field name="name">Sell Rented Product</field>
     <field name="sequence">100</field>
+    <field name="company_id" eval="False" />
```
Same as above

PicchiSeba reviewed (commented):
Review: a few questions here and there

yweng8111 reviewed (approved):
thank you for your contribution
